### PR TITLE
Add --import-file to pulumi preview

### PR DESCRIPTION
**Pull request by Frassle**
# Description

Fixes https://github.com/pulumi/pulumi/issues/12768.

This will generate an import file for every resource the preview wants to Create.

## Checklist

- [x] I have run `make tidy` to update any new dependencies
- [x] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by Frassle** [comment]
>  I have added tests that prove my fix is effective or that my feature works

This probably wants an integration test in tests/smoke_test.go but there are already some tests covering the main code for this.

**Comment by Frassle** [comment]
Need to do some thinking about the Version and PluginDownloadURL options. They aren't on resource events, only on the providers which is fine for explicit providers but we don't see those events for default providers. Might need to change that to fix this. 

Having said that, probably fine to merge this as is then do that as a follow up PR after.

**Comment by Frassle** [comment]
If we can get https://github.com/pulumi/pulumi/pull/14607 in I can rework this to look at the version and url fields on all providers (defaults as well) and then correctly fill in the import file.

**Review by tgummerer — COMMENTED** [review]
I didn't manage to go very in-depth here yet, so I left just a couple of comments on things I noticed while reading through.

File: pkg/cmd/pulumi/import.go [hunk 1]
```diff
@@ -175,7 +175,15 @@ func readImportFile(p string) (importFile, error) {
 	return result, nil
 }
 
-func writeImportFile(v importFile) (string, error) {
+func writeImportFile(v importFile, f io.Writer) error {
+	enc := json.NewEncoder(f)
+	enc.SetEscapeHTML(false)
```
[the remark this answers is not in the document]
Comment (Frassle) [reply pkg/cmd/pulumi/import.go:180]: This is specifically to make the strings safe for HTML, which we really don't care about. The JSON encoder will always escape characters that are needed to make it valid JSON.

File: tests/smoke_test.go [hunk 2]
```diff
@@ -362,3 +362,54 @@ func TestConvertDisableAutomaticPluginAcquisition(t *testing.T) {
...
+	// Make sure random is installed
+	out, _ := e.RunCommand("pulumi", "plugin", "ls")
+	if !strings.Contains(out, "random  resource  4.12.0") {
+		e.RunCommand("pulumi", "plugin", "install", "resource", "random", "4.12.0")
+	}
+
+	e.RunCommand("pulumi", "login", "--cloud-url", e.LocalURL())
+	e.RunCommand("pulumi", "stack", "init", "test")
+	e.RunCommand("pulumi", "install")
+	e.RunCommand("pulumi", "preview", "--import-file", "import.json")
```
[the remark this answers is not in the document]
Comment (Frassle) [reply tests/smoke_test.go:384]: We're going to delete the whole temp directory this is copied into anyway.

File: pkg/cmd/pulumi/preview.go [hunk 3]
```diff
@@ -18,22 +18,152 @@ import (
...
+			// If this is a provider we need to note we've seen it so we can build the Version and PluginDownloadURL of
+			// any resources that use it.
+			if providers.IsProviderType(urn.Type()) {
+				providerInputs[urn] = preEvent.Metadata.Res.Inputs
+			}
+
+			// Only interested in creates
+			if preEvent.Metadata.Op != deploy.OpCreate {
+				continue
+			}
```
Comment (justinvp) [inline pkg/cmd/pulumi/preview.go:74]: So this only works on fresh (empty) stacks? I can't generate an import file based on an existing stack with already created resources?

File: pkg/cmd/pulumi/preview.go [hunk 4]
```diff
@@ -18,22 +18,152 @@ import (
...
+
+			var provider, version, pluginDownloadURL string
+			if new.Provider != "" {
+				ref, err := providers.ParseReference(new.Provider)
+				if err != nil {
+					return importFile{}, fmt.Errorf("could not parse provider reference: %v", err)
+				}
+
+				// If we're trying to create this provider in the same deployment and it's not a default provider then
+				// we need to error, the import system can't yet "import" providers.
```
Comment (justinvp) [inline pkg/cmd/pulumi/preview.go:114]: Is there an issue tracking this?

File: pkg/cmd/pulumi/preview.go [hunk 5]
```diff
@@ -18,22 +18,152 @@ import (
...
+				}
+
+				pluginDownloadURL, err = providers.GetProviderDownloadURL(inputs)
+				if err != nil {
+					return importFile{}, fmt.Errorf("could not get provider download url for %s: %v", ref, err)
+				}
+			}
+
+			var id resource.ID
+			// id only needs filling in for custom resources, set it to a placehold so the user can easily search for that.
```
Comment (justinvp) [inline pkg/cmd/pulumi/preview.go:145]:
```suggestion
			// id only needs filling in for custom resources, set it to a placeholder so the user can easily search for that.
```

File: pkg/cmd/pulumi/preview.go (hunk 4 above)
Comment (justinvp) [reply pkg/cmd/pulumi/preview.go:114]: What does this mean in practice for users that have explicit providers in their stack? The resources that use that explicit provider will be importable? But the explicit providers themselves aren't importable?

**Review by justinvp — APPROVED** [review]
Some questions, but LGTM

File: tests/smoke_test.go [hunk 6]
```diff
@@ -362,3 +362,54 @@ func TestConvertDisableAutomaticPluginAcquisition(t *testing.T) {
...
+	e := ptesting.NewEnvironment(t)
+	defer deleteIfNotFailed(e)
+
+	e.ImportDirectory("testdata/import_node")
+
+	// Make sure random is installed
+	out, _ := e.RunCommand("pulumi", "plugin", "ls")
+	if !strings.Contains(out, "random  resource  4.12.0") {
+		e.RunCommand("pulumi", "plugin", "install", "resource", "random", "4.12.0")
+	}
```
[the remark this answers is not in the document]
Comment (Frassle) [reply tests/smoke_test.go:379]: Ah good point, this was just copying what we do in other tests. I'll fix all of them up.

File: pkg/cmd/pulumi/preview.go [hunk 7]
```diff
@@ -302,6 +464,10 @@ func newPreviewCmd() *cobra.Command {
 	if !hasExperimentalCommands() {
 		contract.AssertNoErrorf(cmd.PersistentFlags().MarkHidden("save-plan"), `Could not mark "save-plan" as hidden`)
 	}
+	cmd.PersistentFlags().StringVar(
+		&importFilePath, "import-file", "",
+		"Save any creates seen during the preview into an import file to use with `pulumi import`")
```
[the remark this answers is not in the document]
Comment (Frassle) [reply pkg/cmd/pulumi/preview.go:469]: Yeh I think this just need good docs, and maybe a comment in the "import" help that "preview --import-file" can build an import file for you.

File: pkg/cmd/pulumi/preview.go (hunk 4 above)
Comment (Frassle) [reply pkg/cmd/pulumi/preview.go:114]: It errors that the provider needs to be created first then they can import it. See a few lines down "cannot import resource %q with a new explicit provider %q".

File: tests/smoke_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Frassle) [reply tests/smoke_test.go:379]: https://github.com/pulumi/pulumi/pull/14723

**Pull request merged**